# Домашнє завдання до теми "Розширені методи обробки даних у Pandas"

В цьому домашньому завданні продовжуємо працювати з набором даних `supermarket_sales.csv`.

0. Імпортуйте бібліотеку pandas та зчитайте дані у змінну `df` типу `pandas.DataFrame`.

In [2]:
import pandas as pd

In [4]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [6]:
data_path = 'drive/MyDrive/Data_analitic/Python/Pandas_Python/supermarket_sales.csv'
df = pd.read_csv(data_path)

1. Відсортуйте дані за спаданням ціни товара ('Unit price') і виведіть 5 перших рядочків на екран.

In [5]:
df.sort_values(by = 'Unit price', ascending=False).head()

Invoice ID Branch       City Customer type Gender         Product line  \
983  148-41-7930      C  Naypyitaw        Normal   Male    Health and beauty   
122  219-22-9386      B   Mandalay        Member   Male    Sports and travel   
930  641-62-7288      B   Mandalay        Normal   Male   Home and lifestyle   
494  437-53-3084      B   Mandalay        Normal   Male  Fashion accessories   
283  667-92-0055      A     Yangon        Member   Male    Health and beauty   

     Unit price  Quantity  Tax 5%    Total       Date   Time      Payment  \
983       99.96         7  34.986  734.706  1/23/2019  10:33         Cash   
122       99.96         9  44.982  944.622   3/9/2019  17:26  Credit card   
930       99.92         6  29.976  629.496  3/24/2019  13:33      Ewallet   
494       99.89         2   9.989  209.769  2/26/2019  11:48      Ewallet   
283       99.83         6  29.949  628.929   3/4/2019  15:02      Ewallet   

       cogs  gross margin percentage  gross income  Rating  
983  699.72                 4.761905        34.986     6.1  
122  899.64                 4.761905        44.982     4.2  
930  599.52                 4.761905        29.976     7.1  
494  199.78                 4.761905         9.989     7.1  
283  598.98                 4.761905        29.949     8.5

2. Знайдіть інвойс (рядок в даних) в категорії товарів `Electronic accessories`, в якому був куплений товар з найбільшою кількістю і найбільшою ціною. Виведіть цей рядок даних на екран.

In [9]:
df[df['Product line'] == 'Electronic accessories'].sort_values(by= ['Unit price', 'Quantity'], ascending= (False, False)).head(1)

Invoice ID Branch      City Customer type  Gender  \
209  817-69-8206      B  Mandalay        Normal  Female   

               Product line  Unit price  Quantity   Tax 5%     Total  \
209  Electronic accessories       99.73         9  44.8785  942.4485   

         Date   Time      Payment    cogs  gross margin percentage  \
209  3/2/2019  19:42  Credit card  897.57                 4.761905   

     gross income  Rating  
209       44.8785     6.5

3.1. Створіть нову колонку в даних `Rating Category`, яка визначає категорію рейтингу (`Rating`): високий, середній чи низький за наступним правилом:
- якщо рейтинг вищий або рівний 8 - то він високий (`High`)
- якщо рейтинг від 6 до 8 не включаючи - то він середній (`Medium`)
- інакше - низький (`Low`).

In [14]:
def rating_category(Rating):
  if Rating >= 8:
    return 'High'
  elif 6 <= Rating < 8:
    return 'Medium'
  else:
    return 'Low'


In [15]:
df['Rating Category'] = df['Rating'].apply(rating_category)

3.2. Виведіть, який відсоток даних становить кожне зі значень категорій рейтингу.

In [16]:
rating_percentages = df['Rating Category'].value_counts(normalize=True) * 100
print(rating_percentages)

Rating Category
Medium    35.6
High      32.9
Low       31.5
Name: proportion, dtype: float64


4.1. Запустіть код нижче аби створити змінну `time`. Ця змінна містить час в форматі "година:хвилина". Напишіть функцію `get_hour`, яка буде зі змінної формату як `time` дістати лише години і конвертувати їх у тип `int`.

Запустіть написану функцію на змінній `time` та перевірте правильність роботи функції.

In [17]:
time = '11:28'

In [18]:
def get_hour(time_str):
    hour = time_str.split(':')[0]
    return int(hour)

print(get_hour(time))

11


4.2. З допомогою функції `get_hour` і метода `DataFrame.apply()` створіть нову колонку `Hour` в нашому наборі даних, яка буде містити години здійснення покупки, обчислені на основі часу здійснення покупки `Time`.

In [21]:
df['Hour'] = df['Time'].apply(get_hour)

In [22]:
print(df[['Time', 'Hour']].head())

    Time  Hour
0  13:08    13
1  10:29    10
2  13:23    13
3  20:33    20
4  10:37    10


4.3. Виведіть кількість значень за кожною годиною покупки і дайте відповідь "О яких годинах найбільше здійснюють покупок покупці і чому на Ваш погляд саме о цих годинах? Чи це слогічно, що найбільше покупок саме в ці години?".

In [24]:
count_h = df['Hour'].value_counts()
print(count_h)

Hour
19    113
13    103
15    102
10    101
18     93
11     90
12     89
14     83
16     77
20     75
17     74
Name: count, dtype: int64


Найбільше покупок здійснюється о 19:00, 13:00, 15:00 та 10:00.
Можливі причини:
19:00 — це час після завершення робочого дня. Люди повертаються додому, заходять до супермаркетів або роблять онлайн-замовлення продуктів на вечерю чи наступний день. Це логічно, бо саме ввечері з’являється вільний час для покупок.
13:00 та 15:00 — обідній період. У більшості компаній це час перерви, і багато людей купують їжу на обід або заздалегідь для вечора, щоб не витрачати час після роботи.
10:00 — ранковий пік. Люди, які не встигли зробити покупки зранку або хочуть підготуватись до робочого дня (наприклад, купити обід чи продукти додому), роблять це саме в цей час.

На мій погляд, це цілком логічно і відповідає поведінці покупців у реальному житті. Розподіл покупок свідчить про прив’язку до розпорядку дня:
перед роботою,
під час обідньої перерви,
після завершення робочого дня.

5. Виведіть середнє значення чеку (`Total`) для покупок в містах, наявних в даних, за допомогою функції `groupby`.

In [6]:
df.groupby('City')['Total'].mean()

City
Mandalay     319.872506
Naypyitaw    337.099715
Yangon       312.354031
Name: Total, dtype: float64

6. Виведіть кількість покупок, середнє, мінімальне та максимальне значення для загальної суми чеку (Total) в містах, наявних в даних, за допомогою функції groupby.
Зробіть висновки з цих показників та дайте рекомендації для мережі супермаркетів.

In [10]:
df.groupby('City')['Total'].agg(['count', 'mean', 'min', 'max']).round(2).sort_values(by='mean', ascending=False)

count    mean    min      max
City                                    
Naypyitaw    328  337.10  10.68  1042.65
Mandalay     332  319.87  18.64  1022.49
Yangon       340  312.35  12.69  1039.29

За кількістю покупок лідером є Yangon — 340 транзакцій. Це свідчить про найвищу загальну активність покупців у цьому місті.
Найвищий середній чек зафіксовано в Naypyitaw — 337.10. Також саме тут була здійснена найдорожча покупка — 1042.65, що свідчить про наявність платоспроможних клієнтів або великі закупівлі.
Найнижчий чек також у Naypyitaw — 10.68. Це може свідчити про наявність дрібних покупок, акційних товарів або тестових транзакцій. Mandalay має середні показники по всіх категоріях, але вищій мінімальний чек. Це може свідчити про стабільні середні витрати та менш виражений економ-сегмент.

Той факт, що в одному місті (Naypyitaw) зафіксовано як найвищий, так і найнижчий чек, свідчить про велике різноманіття типів покупців — від економних до преміальних. Необхідно провести аудит асортименту у Mandalay та Yangon — можливо, тут бракує товарів середнього/високого класу.Звернути увагу на якість обслуговуання, чи якість продуктів. Використовувати різні маркетингові підходи для різних типів міст і покупців. Збирати додаткові дані про покупців для персоналізації акцій та підвищення лояльності.

7. Ми внесли деякі зміни в наш набір даних, тож, давайте запишемо результат у файл. Запишіть оновлену версію даних у файл `supermarket_sales_enriched.csv`. Зверніть увагу, аби не записати нічого зайвого, наприклад, індекси рядків нам в файлі не потрібні, оскільки вони стандартні.

In [7]:
df.to_csv('supermarket_sales_enriched.csv', index=False)